### ISSUE TITLE
ORM

### ISSUE THREAD

**Issue by nviennot** [kilt-cms/kilt-cms#13]
Hi guys,

I was wondering if you guys had though of using http://nobrainer.io/ as I would think NoBrainer would be a good fit for this project when I see this: https://github.com/AsheAvenue/Kilt/blob/master/lib/kilt.rb

If you did, but NoBrainer wasn't helpful for you, what were the missing features?

**Comment by timboisvert** [comment]
Nicolas: I'm familiar with Nobrainer and we've used it in a few POCs, but using an ORM wasn't appropriate in this case because we don't have statically-defined classes/objects. Our "models" are essentially inferred at runtime based on what's defined in the Kilt config file, and the config can change without having to reload the app. Since all we ever load is our "Object" class, we don't have a need for an ORM. Additionally, all objects in Kilt share a common table in Rethink and differentiate their type based solely on a single string value. 

However, I will say that when we do have a Rethink-based project that does have multiple object types and we want to use an ORM, we'll absolutely use Nobrainer.

**Comment by nviennot** [comment]
I see :) I would still argue that it would be useful to use nobrainer (either with dynamic attributes, or defining fields on the fly at runtime by calling `KiltObject.field :some_field, :type => String` for example). But I guess I'm a bit _too_ eager to see nobrainer in action, and perhaps it's just not a good use case :)

I'm glad to hear that you would consider it for other projects though, because I'm dying to get some users :)

Thanks,
Nico.

**Comment by darrencauthon** [comment]
NoBrainer seems like a good library, but I'd like to argue against using it for another reason:

By relying on core Ruby objects, Kilt is on the cusp of being able to support any database... or even support **NOT** having a database (pstore?).  The data is JSON, so anything that can store JSON could work.

I've abstracted out the database work that I've seen from Kilt here (https://github.com/darrencauthon/Kilt/blob/777873bdba73cd3869c48f5f60af80cf886cf6e3/lib/kilt/database.rb).  It would be trivial to write another class to implement these features.

I know that Kilt was not meant to be database-agnostic, but... gosh, it's so close!  Using NoBrainer would kill this.

**Comment by nviennot** [comment]
I see, well not really. I understand the idea of being database agnostic, but I don't understand the benefits. Specifically, I don't understand why it's a good thing to support any database as they often provide different semantics. Switching databases once the application is deployed is something that I haven't seen often, except perhaps during a heavy application refactor.
I would think that being database agnostic prevents a full range of database optimizations that could be possible otherwise, and starts becoming a burden for the developer. Even simple tasks such as doing the simplest database migration become a hurdle. That's without mentioning that some migrations may require the use of atomic operations, which makes it a little hard to stay database agnostic.

Having said that, I understand the intellectual satisfaction of such an abstraction. This reminds me of the approach of DataMapper: http://datamapper.org/why.html and now Ruby Object Mapper https://github.com/rom-rb/rom that tries to abstract the datastore as much as possible, while keeping a powerful API.
Then there is orm_adapter which was extracted from Devise: https://github.com/ianwhite/orm_adapter which tries to put an ORM abstraction on top of existing ORMs.
Check out these projects if the topic interests you :)

Nico.

**Comment by darrencauthon** [comment]
There's no intellectual satisfaction in this abstraction.  There's a concrete one:  I'm launching a couple simple, content-based app by May, and my main hosting options are Heroku and Engine Yard.  Neither come with support for RethinkDB (though Heroku has a "beta" RethinkDB add-on), but both come with Postgres databases with automated backups.  I _could_ set up RethinkDB separately from the app, but why?  And where do the new, non-CRUD features that I add to the app  separate from Kilt go (like possibly Devise, or search indexing)... in the Postgres database, or RethinkDB?  With the former, I have two databases to maintain and confuse devs, and with the latter I'm going to try to fit the square Rails peg (relational-table db) into a round hole (document db).  

A month or two ago, my opinion was to stick with RethinkDB, as Kilt was built for RethinkDB.  However, when I got into the guts of Kilt, I saw that the RethinkDB use was relatively limited.  By simply copy-pasting the database interactions into a separate class, we could create a contract that actually relies on very little from RethinkDB.  Look at the code I linked above... you'll see a handful of simple CRUD operations.  Operations so simple, any database's "full range of database optimizations" will do.

I think one of the benefits of Kilt is that you're still free to implement your own front-end, which includes hitting the database with whatever tool you wish.  I could use Kilt today, as-is, and then use NoBrainer on the front-end.  Kilt doesn't take over my app like other Rails CRMs... it just provides me 95% of what I used those CRMs for, but with 5% of the code.  I hope Kilt never grows to the point that it truly depends on specifics of a database, whether it's RethinkDB or not.  Once it crosses that line, it'll be just like the others.

But if the topic of database-abstraction interests you, Nico, here's a fun talk by Uncle Bob that might have some relation to Kilt.  I can't remember the full story, but I remember him talking about a talk where devs were building an app (either Fit or a wiki??).  Their initial thought was to start by picking a database, but instead they kept putting off that decision while building an app.  They kept building and building, and it wasn't until much later that they realized... they didn't need a database at all!

http://www.confreaks.com/videos/759-rubymidwest2011-keynote-architecture-the-lost-years
